frontend/packages/topology/src/layouts/DagreLayout.ts

```diff
@@ -1,119 +1,142 @@
-  }
+class DagreGroup extends LayoutGroup {}
 
-  get x(): number {
-    return this.node.getBounds().getCenter().x;
-  }
+class DagreLink extends LayoutLink {
+  public points: any[];
 
-  set x(value: number) {
-    this.node.setBounds(
-      this.node.getBounds().setCenter(value, this.node.getBounds().getCenter().y),
-    );
+  updateBendpoints(): void {
+    if (this.points && !this.isFalse && this.points.length > 2) {
+      this.element.setBendpoints(
+        this.points.slice(1, -1).map((point: any) => new Point(point.x, point.y)),
+      );
+    }
   }
+}
 
-  get y(): number {
-    return this.node.getBounds().getCenter().y;
+type DagreLayoutOptions = LayoutOptions & {
+  marginx: number;
+  marginy: number;
+  nodesep: number;
+  edgesep: number;
+  ranker: string;
+  rankdir: string;
+};
+
+class DagreLayout extends BaseLayout implements Layout {
+  private dagreOptions: DagreLayoutOptions;
+
+  constructor(graph: Graph, options?: Partial<DagreLayoutOptions>) {
+    super(graph, options);
+    this.dagreOptions = {
+      ...this.options,
+      ...{
+        marginx: 0,
+        marginy: 0,
+        nodesep: this.options.nodeDistance,
+        edgesep: this.options.linkDistance,
+        ranker: 'tight-tree',
+        rankdir: 'TB',
+      },
+      ...options,
+    };
   }
 
-  set y(value: number) {
-    this.node.setBounds(
-      this.node
-        .getBounds()
-        .clone()
-        .setCenter(this.node.getBounds().getCenter().x, value),
-    );
+  protected createLayoutNode(node: Node, nodeDistance: number, index: number) {
+    return new DagreNode(node, nodeDistance, index);
   }
-}
-
-class DagreEdge {
-  public points: any[];
 
-  private edge: Edge;
-
-  constructor(edge: Edge) {
-    this.edge = edge;
+  protected createLayoutLink(
+    edge: Edge,
+    source: LayoutNode,
+    target: LayoutNode,
+    isFalse: boolean,
+  ): LayoutLink {
+    return new DagreLink(edge, source, target, isFalse);
   }
 
-  get element(): Edge {
-    return this.edge;
+  protected createLayoutGroup(node: Node, padding: number, index: number) {
+    return new DagreGroup(node, padding, index);
   }
 
-  get source(): string {
-    return this.edge.getSource().getId();
+  protected updateEdgeBendpoints(edges: DagreLink[]): void {
+    _.forEach(edges, (edge) => {
+      const link = edge as DagreLink;
+      link.updateBendpoints();
+    });
   }
 
-  get target(): string {
-    return this.edge.getTarget().getId();
+  protected startLayout(graph: Graph, initialRun: boolean, addingNodes: boolean): Promise<any> {
+    if (initialRun || addingNodes) {
+      const dagreGraph = new dagre.graphlib.Graph({ compound: true });
+      dagreGraph.setGraph({
+        marginx: this.dagreOptions.marginx,
+        marginy: this.dagreOptions.marginy,
+        nodesep: this.dagreOptions.nodesep,
+        edgesep: this.dagreOptions.edgesep,
+        ranker: this.dagreOptions.ranker,
+        rankdir: this.dagreOptions.rankdir,
+      });
```

Comment:
I think we need a way to apply all options I send you as long as it's valid Dagre options. I'm using `align` and `ranksep` for instance, to do my layout.